### PR TITLE
Web test

### PR DESCRIPTION
**Pull request by andresmadeddie**
Install selenium server and Chromedriver
Configure config.properties
Create three Page Object classes Login, Product, and Cart
Create one component class
Create one util class for authentication
Create one abstract class
Add default username and password in testdata.properties
Create SauceDemoTest Class
Test successful
Add Web Test Cases file on the README.md

### CODE REVIEW

File: src/main/java/com/solvd/carina/demo/gui/pages/saucedemo/CartPage.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    private ExtendedWebElement pageTitleLabel;
+
+    @FindBy(xpath = "//div[@class='inventory_item_name'][text()='%s']")
+    private ExtendedWebElement productTitle;
+
+    protected CartPage(WebDriver driver) {
+        super(driver);
+        setPageOpeningStrategy(PageOpeningStrategy.BY_ELEMENT);
+        setUiLoadedMarker(pageTitleLabel);
+//        setPageAbsoluteURL(R.CONFIG.get("url"));
```
Comment (ankochergin) [inline src/main/java/com/solvd/carina/demo/gui/pages/saucedemo/CartPage.java]: please remove if not needed

File: src/main/java/com/solvd/carina/demo/gui/pages/saucedemo/ProductsPage.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    private ExtendedWebElement cartButton;
+
+    @FindBy(xpath = "//div[@class='inventory_item']")
+    private List<CatalogProductItem> products;
+
+    protected ProductsPage(WebDriver driver) {
+        super(driver);
+        setPageOpeningStrategy(PageOpeningStrategy.BY_ELEMENT);
+        setUiLoadedMarker(pageTitleLabel);
+//        setPageAbsoluteURL(R.CONFIG.get("url"));
```
Comment (ankochergin) [inline src/main/java/com/solvd/carina/demo/gui/pages/saucedemo/ProductsPage.java]: please remove if not needed

File: src/test/java/com/solvd/carina/demo/SauceDemoTest.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+import java.util.List;
+
+public class SauceDemoTest extends AbstractSauceDemoTest {
+
+    @Test
+    @MethodOwner(owner = "Andres")
+    public void LoginTest() {
```
Comment (ankochergin) [inline src/test/java/com/solvd/carina/demo/SauceDemoTest.java]:
```suggestion
    public void loginTest() {
```

File: src/test/java/com/solvd/carina/demo/SauceDemoTest.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        loginPage.typeUsername(username);
+        loginPage.typePassword(password);
+        ProductsPage productsPage = loginPage.clickLoginButton();
+        Assert.assertTrue(productsPage.isPageOpened(), "product page is not opened");
+    }
+
+    @Test
+    @MethodOwner(owner = "Andres")
+    public void productAddedToTheCartTest() {
+        String productTitle = "Sauce Labs Onesie";
```
Comment (jinocencio-solvd) [inline src/test/java/com/solvd/carina/demo/SauceDemoTest.java]: maybe use dataProvider?

File: src/main/java/com/solvd/carina/demo/gui/saucedemo/ProductsPage.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+    public LeftSideMenu getLeftSideMenu() {
+        return leftSideMenu;
+    }
+
+    public List<CatalogProductItem> getProductItems() {
+        return products;
+    }
+
+    public CartPage clickCartButton() {
```
Comment (jinocencio-solvd) [inline src/main/java/com/solvd/carina/demo/gui/saucedemo/ProductsPage.java:53]: i think this method name could be more descriptive like openCartPage or getCartPage

File: src/test/java/com/solvd/carina/demo/TestWithThreads.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
+package com.solvd.carina.demo;
+
+import java.util.concurrent.*;
+
+public class TestWithThreads {
```
Comment (jinocencio-solvd) [inline src/test/java/com/solvd/carina/demo/TestWithThreads.java]: Have you tried configuring Maven/Testng or Carina configs to run methods  in parallel?

File: src/test/java/com/solvd/carina/demo/SauceDemoTest.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        String username = R.TESTDATA.get("standard_username");
+        String password = R.TESTDATA.get("standard_password");
+
+        LoginPage loginPage = new LoginPage(getDriver());
+        loginPage.open();
+        Assert.assertTrue(loginPage.isPageOpened(), "Login page is not opened");
+        loginPage.typeUsername(username);
+        loginPage.typePassword(password);
+        ProductsPage productsPage = loginPage.clickLoginButton();
+        Assert.assertTrue(productsPage.isPageOpened(), "product page is not opened");
```
Comment (jinocencio-solvd) [inline src/test/java/com/solvd/carina/demo/SauceDemoTest.java]: there is also assertPageOpened method accessible within AbstractPage class

File: README.md [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+This repository contains automated test cases for the "Sauce Demo" website. The test cases are written using Selenium and are designed to check the functionality of the website's login process and the "add to cart" feature.
+
+Note: Can use the next script to run the Selenium server before the test execution 
+
+```
+java -jar selenium-server-4.10.0.jar standalone
+```
+## Test Cases
+
+| Feature Owner (Author) | #   | Name                   | Preconditions                                                                                                  | Step # | Description                                            | Expected Result        |
```
Comment (jinocencio-solvd) [inline README.md:14]: Nice job with the table display!

File: src/main/java/com/solvd/carina/demo/gui/saucedemo/components/LeftSideMenu.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    private ExtendedWebElement logOutButtonInLeftSideMenu;
+
+    public LeftSideMenu(WebDriver driver, SearchContext searchContext) {
+        super(driver, searchContext);
+    }
+
+    public LoginPage clickLogOutButtonInLeftSideMenu() {
+        logOutButtonInLeftSideMenu.click();
+        return new LoginPage(driver);
+    }
```
Comment (ankochergin) [inline src/main/java/com/solvd/carina/demo/gui/saucedemo/components/LeftSideMenu.java]: it's already in LeftSideMenu so it's not needed to write so long names for methods and variables
```suggestion
    public LoginPage clickLogOutButton() {
        logOutButton.click();
        return new LoginPage(driver);
    }
```

File: src/main/java/com/solvd/carina/demo/gui/saucedemo/components/LeftSideMenu.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+    public AboutPage clickAboutButtonInLeftSideMenu() {
+        aboutButtonInLeftSideMenu.click();
+        return new AboutPage(driver);
+    }
```
Comment (ankochergin) [inline src/main/java/com/solvd/carina/demo/gui/saucedemo/components/LeftSideMenu.java]: same as in previous comment

File: src/test/java/com/solvd/carina/demo/SauceDemoTest.java [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+    @Test
+    @MethodOwner(owner = "Andres")
+    public void loginTest() {
+        String username = R.TESTDATA.get("standard_username");
+        String password = R.TESTDATA.get("standard_password");
+
+        LoginPage loginPage = new LoginPage(getDriver());
+        loginPage.open();
+        loginPage.assertPageOpened();
```
Comment (ankochergin) [inline src/test/java/com/solvd/carina/demo/SauceDemoTest.java:23]: I prefer to write assertTrue(page.isPageOpened(), "XXX page is not opened");
It helps to customise message and code is more readable

File: src/test/resources/testng_suites/web.xml [hunk 12]
```diff
@@ -17,6 +17,7 @@
 	<test name="GSM arena web tests">
 		<classes>
 			<class name="com.solvd.carina.demo.WebSampleTest" />
+			<class name="com.solvd.carina.demo.SauceDemoTest"/>
```
Comment (ankochergin) [inline src/test/resources/testng_suites/web.xml]: specify each test method individually using <include> node to make it easier to support

**Pull request merged**